### ISSUE TITLE
Switching CPU

### ISSUE THREAD

**Issue by huydinhtran** [bgodala/QPoints#1]
Hi,

I'm trying to implement a switch CPU. Since the gem5 version that you are using is quite old, how can I implement switching CPU? I tried using SimpleSwitchableCPU() and m5.switchCpus but had no luck. It's probably since the gem5 version is old that it was not supported for ARM back then. 

Any help would be greatly appreciated,
Thanks!

**Comment by bgodala** [comment]
@huydinhtran Switchable CPUs used to work in the older version of gem5 as well. If you can share more details on what is happening would be helpful to understand the underlying issue. Please outputs and configs, hopefully that would be enough to understand where is the issue.

**Comment by huydinhtran** [comment]
I'm not too sure how to implement SwitchableCPU() on the starter_fs.py script that you used for QPoints. Since that script is modified to work with QEMU checkpoints, the way it sets up the system is quite complicated. So I'm not too sure how I can implement switching inside of the devices.py since it is used for setting up the system for starter_fs.py.

**Comment by huydinhtran** [comment]
Could you share the overall method that you implemented to make gem5 work with QEMU checkpoints? I would like to implement QPoints for the current version of gem5.

**Comment by bgodala** [comment]
Yes, but I may not be able to explain the whole process here. I presented the overall method at gem5 workshop. Here is the link to the poster (https://www.gem5.org/assets/files/workshop-isca-2023/posters/qpoints.pdf). I hope that helps as a starting point.
At a high level there are three key components.
1. QEMU Virt Config in gem5 ( GEM5_HOME/src/dev/arm/RealView.py, look for QEMU_Virt class)
2. Checkpoint template (QPoints/scripts/templates, add a new template file compatible with the newer version of gem5)
3. Raw data to extract from QEMU emulated system (QPoints/scripts/create_snapshot.py)

In order to port QPoints  to newer version of gem5 all three components probably need to be updated. I hope this helps you get started with porting.

**Comment by huydinhtran** [comment]
From the devices.py and starter_fs.py scripts, besides the QEMU_Virt, what are some of the important modifications that must be included to run QEMU checkpoints? 

Can I just run the QEMU checkpoints by creating another simple script that uses AbstractBoard and AbstractProcessor or do I have to use the starter_fs.py script?

**Comment by bgodala** [comment]
I'm not completely familiar with newer gem5 infrastructure. A generic checkpoint script that works with gem5 should also work with QEMU checkpoints if that script (AbstractBoard or AbstractProcessor) uses QEMU_Virt configuration.

I remember fixing some bugs to support multi-core QEMU checkpoint. I don't remember changes but it can be found by looking at commit log messages. Those messages probably contains words "multicore", "gic" or "m1".

**Comment by huydinhtran** [comment]
For the current gem5, the GenericTimer of ARM got "renamed"/modified, and maybe other system updates as well which preventing me from porting QPoints to the current gem5. I tried to implement QEMU_Virt for the current gem5, and tested with the new and old GenericTimer, but wasn't able to restore the QEMU checkpoints. This is the error message when trying to run the simulation with current gem5:
``gem5.opt: src/sim/eventq.hh:816: void gem5::EventQueue::reschedule(gem5::Event*, gem5::Tick, bool): Assertion `when >= getCurTick()' failed.``

I talked to several gem5 devs and looks like it might not be possible to restore old checkpoints in current gem5 given the fact that QPoints was based on a 2-year-ago version of gem5.  

I guess I might need to look into the checkpoint that was created to be compatible with the current gem5. Could you share how were you able to create the checkpoint that is compatible with gem5? Is there an exclusive section within the creating checkpoint process that is specific to a version of gem5?

**Comment by bgodala** [comment]
I created a sample checkpoint using gem5's Atomic CPU Model to understand various fields gme5 needs to resume it. Then I created template file using that just to plug in architectural state values like CPU registers, misc registers and disk controller state (this stuff is anyway automated now).  There are several things that needs to match between two systems. The hardware mapping must be exactly same. So, I created a QEMU_virt config to match QEMU's virt board. I think I added some fixes to Generic Interrupt Controller code to support apple's M1 Mac. I have also added some fixes to enable multi-core system.
